systemtest/infra_kafka.go

```diff
@@ -0,0 +1,230 @@
+)
+
+var brokersMu sync.RWMutex
+var kafkaBrokers = []string{"localhost:9093"}
+
+var (
+	runOnce      sync.Once
+	mockResolver *net.Resolver
+)
+
+// KafkaConfig for Kafka cluster provisioning with topics.
+type KafkaConfig struct {
+	// TFPath is the path where the terraform files are present.
+	TFPath string
+	// Namespace to pass down as the `namespace` Terraform var.
+	Namespace string
+	// Namespace to pass down as the `name` Terraform var.
+	Name string
+	// Topics to create in the Kafka cluster. It's passed down as the
+	// `topics` Terraform var.
+	Topics []string
+	// PortForward. If set to true, it forwards traffic to PortForwardResource
+	// using PortForwardMapping.
+	PortForward bool
+	// PortForwardResource specifies the resource to which the traffic
+	// is forwarded.
+	PortForwardResource string
+	// PortForwardMapping holds the port forward mapping for the specified
+	// resource.
+	PortForwardMapping string
+}
+
+func newLocalKafkaConfig(topics ...string) KafkaConfig {
+	return KafkaConfig{
+		Topics:      topics,
+		PortForward: true,
+		TFPath:      filepath.Join("tf", "kafka"),
+	}
+}
+
+// ProvisionKafka provisions a Kafka cluster. The specified terraform path
+// must accept at least these variables:
+// - namespace (string)
+// - name (string)
+// - topics (list(string))
+// The terraform module should output at least these variables:
+// - deployment_type (string)
+// - kafka_brokers (list(string))
+//
+// If `deployment_type` output is "k8s", `kubectl port-forward` will be run
+// targeting the `PortForwardResource` with `PortForwardMapping`.
+func ProvisionKafka(ctx context.Context, cfg KafkaConfig) error {
+	if cfg.Name == "" {
+		if n := os.Getenv("KAFKA_NAME"); n != "" {
+			cfg.Name = n
+		} else {
+			cfg.Name = "kafka"
+		}
+	}
+	if cfg.Namespace == "" {
+		if n := os.Getenv("KAFKA_NAMESPACE"); n != "" {
+			cfg.Namespace = n
+		} else {
+			cfg.Namespace = "kafka"
+		}
+	}
+	logger().Infof("provisioning Kafka infrastructure with config: %+v", cfg)
+	tf, err := NewTerraform(ctx, cfg.TFPath)
+	if err != nil {
+		return err
+	}
+	if err := tf.Init(ctx, tfexec.Upgrade(true)); err != nil {
+		return err
+	}
+	// Create a dummy topic so the infrastructure awaits until all the required
+	// infrastructure is available. It will be destroyed during the first test.
+	if len(cfg.Topics) == 0 {
+		cfg.Topics = append(cfg.Topics, SuffixTopics("dummy")...)
+	}
+	jsonTopics, err := json.Marshal(cfg.Topics)
+	if err != nil {
+		return err
+	}
+	namespaceVar := tfexec.Var(fmt.Sprintf("namespace=%s", cfg.Namespace))
+	nameVar := tfexec.Var(fmt.Sprintf("name=%s", cfg.Name))
+	topicsVar := tfexec.Var(fmt.Sprintf("topics=%s", jsonTopics))
+	// Ensure terraform destroy runs once per TF path.
+	RegisterDestroy(cfg.TFPath, func() {
+		logger().Info("destroying provisioned Kafka infrastructure...")
+		tf.Destroy(ctx, topicsVar, namespaceVar, nameVar)
+	})
+	if err := tf.Apply(ctx, topicsVar, namespaceVar, nameVar); err != nil {
```

Comment:
**issue**: we should run `tf.Validate` before applying to make sure there are no issues with the file

Reply:
same comment as above